src/main/java/htsjdk/samtools/cram/structure/SliceBlocksWriteStreams.java

```diff
@@ -0,0 +1,117 @@
+import htsjdk.samtools.cram.io.DefaultBitOutputStream;
+import htsjdk.samtools.cram.structure.block.Block;
+import htsjdk.samtools.util.RuntimeIOException;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Provides a layer over a {@link SliceBlocks} object and acts as a bridge between the DataSeries codecs
+ * and their underlying blocks when writing a CRAM stream by presenting a bit (core) or byte (external) stream
+ * for each block.
+ */
+public class SliceBlocksWriteStreams {
+
+    private final CompressionHeader compressionHeader;
+    private final SliceBlocks sliceBlocks;
+
+    private final ByteArrayOutputStream coreBlockByteOutputStream;
+    private final BitOutputStream coreBlockBitOutputStream;
+
+    // content ID to ByteArrayOutputStream
+    private final Map<Integer, ByteArrayOutputStream> externalOutputStreams = new HashMap<>();
+
+    /**
+     * @param compressionHeader {@link CompressionHeader} for the container containing the slice
+     * @param sliceBlocks {@link SliceBlocks} for the represented by these blocks {@link Slice}
+     */
+    public SliceBlocksWriteStreams(final CompressionHeader compressionHeader, final SliceBlocks sliceBlocks) {
+        this.compressionHeader = compressionHeader;
+        this.sliceBlocks = sliceBlocks;
+
+        coreBlockByteOutputStream = new ByteArrayOutputStream();
+        coreBlockBitOutputStream = new DefaultBitOutputStream(coreBlockByteOutputStream);
+
+        // Create an output stream for each external content ID in the encoding map
+        for (final Integer contentID : compressionHeader.getEncodingMap().getExternalIDs()) {
+            externalOutputStreams.put(contentID, new ByteArrayOutputStream());
+        }
+    }
+
+    /**
+     * @return the {@link BitOutputStream}  for the core block
+     */
+    public BitOutputStream getCoreOutputStream() { return coreBlockBitOutputStream; }
+
+    /**
+     * Get the ByteArrayOutputStream corresponding to the requested contentID
+     * @param contentID ID of content being requested
+     * @return ByteArrayOutputStream for contentID
+     */
+    public ByteArrayOutputStream getExternalOutputStream(final Integer contentID) { return externalOutputStreams.get(contentID); }
+
+    /**
+     * Compress and write each each stream to a corresponding Block (note that this does not write
+     * the blocks themselves to a container output stream - that can't happen until the slice is aggregated
+     * into a container.
+     */
+    public void flushStreamsToBlocks() {
+         closeAllStreams();
+
+         // core block is raw (no compression) and must be written first (prescribed by the spec)
+         sliceBlocks.setCoreBlock(Block.createRawCoreDataBlock(coreBlockByteOutputStream.toByteArray()));
+
+         for (final Map.Entry<Integer, ByteArrayOutputStream> streamEntry : externalOutputStreams.entrySet()) {
```

Comment:
I always think map.foreach is nicer than for(Map.entry) but that's a style thing so up to you...